File: solutions/ruby/allergies/1/allergies.rb
class Allergies
  # Definimos los alérgenos en orden de su valor binario
  ALLERGENS = {
    'eggs'         => 1,
    'peanuts'      => 2,
    'shellfish'    => 4,
    'strawberries' => 8,
    'tomatoes'     => 16,
    'chocolate'    => 32,
    'pollen'       => 64,
    'cats'         => 128
  }.freeze

  def initialize(score)
    @score = score
  end

  # Verifica si es alérgico a un item específico
  def allergic_to?(item)
    # Usamos el operador & para ver si el bit del alérgeno está "encendido" en la puntuación
    (@score & ALLERGENS[item]) != 0
  end

  # Devuelve la lista completa de alergias
  def list
    # Filtramos el hash buscando solo las llaves donde allergic_to? sea cierto
    ALLERGENS.keys.select { |item| allergic_to?(item) }
  end
end

File: solutions/typescript/allergies/1/allergies.ts
export class Allergies {
  private score: number;
  private readonly allergens: Record<number, string> = {
    1: 'eggs',
    2: 'peanuts',
    4: 'shellfish',
    8: 'strawberries',
    16: 'tomatoes',
    32: 'chocolate',
    64: 'pollen',
    128: 'cats'
  };

  constructor(allergenIndex: number) {
    this.score = allergenIndex;
  }

  public list(): string[] {
    const allergies: string[] = [];
    
    for (const [value, allergen] of Object.entries(this.allergens)) {
      if (this.isAllergicTo(parseInt(value))) {
        allergies.push(allergen);
      }
    }
    
    return allergies;
  }

  public allergicTo(allergen: string): boolean {
    for (const [value, name] of Object.entries(this.allergens)) {
      if (name === allergen) {
        return this.isAllergicTo(parseInt(value));
      }
    }
    return false;
  }

  private isAllergicTo(value: number): boolean {
    return (this.score & value) === value;
  }
}